File: cpp_backup/pfm/stipple_extras.cpp
#include "pfm/stipple_extras.h"
#include "core/geometry.h"
#include <opencv2/imgproc.hpp>
#include <cmath>

BaseStipplePFM::BaseStipplePFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}

QVector<PFMSetting> BaseStipplePFM::defineSettings() const {
    return {
        {"num_shapes", "Number of Shapes", SettingType::Integer, 5000, QVariant(), 100, 50000, 100, 50000, 100},
        {"shape_size", "Shape Size", SettingType::Number, 1.0, QVariant(), 0.1, 10.0, 0.1, 10.0, 0.1}
    };
}

QVector<DrawingGeometry> BaseStipplePFM::_process(const cv::Mat& image) {
    int w = image.cols;
    int h = image.rows;
    int num_shapes = m_settings["num_shapes"].toInt();
    double size = m_settings["shape_size"].toDouble();

    std::vector<double> probs(w * h, 0.0);
    double sum = 0.0;
    for (int y = 0; y < h; ++y) {
        const uchar* row = image.ptr<uchar>(y);
        for (int x = 0; x < w; ++x) {
            double d = 255.0 - row[x];
            if (d < 0) d = 0;
            probs[y * w + x] = d;
            sum += d;
        }
    }
    if (sum < 1e-6) return {};
    for (auto& p : probs) p /= sum;

    QVector<DrawingGeometry> geoms;
    for (int i = 0; i < num_shapes; ++i) {
        if (isCancelled()) break;
        if (i % 500 == 0) emitProgress(float(i) / num_shapes, geoms.size(), "Stippling...");

        int idx = weightedChoice(probs);
        int cx = idx % w;
        int cy = idx / w;
        double local = probs[idx] * sum / 255.0; // recover brightness
        double r = size * (0.3 + 0.7 * local);

        DrawingGeometry dg;
        dg.path = generateShape(cx, cy, r);
        geoms.push_back(dg);
    }
    return geoms;
}

Path StippleCirclesPFM::generateShape(float cx, float cy, float r) { return generate_circle(cx, cy, r, 12); }
Path StippleSquaresPFM::generateShape(float cx, float cy, float r) { return generate_polygon(cx, cy, r, 4, 45); }
Path StippleTrianglesPFM::generateShape(float cx, float cy, float r) { return generate_polygon(cx, cy, r, 3, 30); }
Path StippleHexagonsPFM::generateShape(float cx, float cy, float r) { return generate_polygon(cx, cy, r, 6, 0); }
Path StippleStarsPFM::generateShape(float cx, float cy, float r) { 
    Path p1 = generate_polygon(cx, cy, r, 5, 0);
    Path p2 = generate_polygon(cx, cy, r*0.5f, 5, 36);
    p1.insert(p1.end(), p2.begin(), p2.end());
    return p1;
}
Path StippleCrossesPFM::generateShape(float cx, float cy, float r) {
    return {{cx-r, cy}, {cx+r, cy}, {cx, cy}, {cx, cy-r}, {cx, cy+r}};
}
Path StippleLinesPFM::generateShape(float cx, float cy, float r) {
    return {{cx-r, cy-r}, {cx+r, cy+r}};
}
Path StippleVariableCirclesPFM::generateShape(float cx, float cy, float r) {
    return generate_circle(cx, cy, r * (0.5 + randUniform(0, 1)), 8);
}
Path StippleVariableSquaresPFM::generateShape(float cx, float cy, float r) {
    return generate_polygon(cx, cy, r * (0.5 + randUniform(0, 1)), 4, randUniform(0, 90));
}
Path StippleChaosPFM::generateShape(float cx, float cy, float r) {
    int pts = int(randUniform(0, 4)) + 3;
    return generate_polygon(cx, cy, r, pts, randUniform(0, 360));
}

File: python/app/pfm/stipple_extras.py
"""Extra Stipple PFMs for DrawingBotV3."""

from typing import List, Callable
import numpy as np
from app.pfm import PathFindingModule, PFMSetting, SettingType, DrawingGeometry, make_random_seed_setting
from app.core.geometry import generate_circle, generate_polygon

class BaseStipplePFM(PathFindingModule):
    @property
    def category(self) -> str: return "Stipple"

    @property
    def is_premium(self) -> bool:
        return True
    
    def _define_settings(self) -> List[PFMSetting]:
        return [
            make_random_seed_setting(),
            PFMSetting("num_shapes", "Number of Shapes", SettingType.INTEGER, 5000, 100, 50000, 100),
            PFMSetting("shape_size", "Shape Size", SettingType.NUMBER, 1.0, 0.1, 10.0, 0.1),
        ]

    def _generate_shape(self, cx: float, cy: float, r: float) -> List[tuple[float, float]]:
        return generate_circle(cx, cy, r, 6) # Default

    def _process(self, image: np.ndarray, progress: Callable) -> List[DrawingGeometry]:
        h, w = image.shape[:2]
        num_shapes = self.get("num_shapes")
        size = self.get("shape_size")
        
        geometries = []
        rng = self._rng
        
        darkness = 255.0 - image.astype(np.float32)
        darkness = np.clip(darkness, 0, 255)
        d_sum = darkness.sum()
        if d_sum < 1.0: return []
        
        probs = darkness.ravel() / d_sum
        indices = rng.choices(range(len(probs)), weights=probs, k=num_shapes)
        
        for i, idx in enumerate(indices):
            if self.is_cancelled: break
            cy, cx = idx // w, idx % w
            local = darkness[cy, cx] / 255.0
            r = size * (0.3 + 0.7 * local)
            path = self._generate_shape(cx, cy, r)
            geometries.append(DrawingGeometry(path=path))
            if i % 500 == 0: progress(i / num_shapes, len(geometries), f"Stippling...")
                
        return geometries


class StippleCirclesPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Circles"
    def _generate_shape(self, cx: float, cy: float, r: float): return generate_circle(cx, cy, r, 12)

class StippleSquaresPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Squares"
    def _generate_shape(self, cx: float, cy: float, r: float): return generate_polygon(cx, cy, r, 4, 45)

class StippleTrianglesPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Triangles"
    def _generate_shape(self, cx: float, cy: float, r: float): return generate_polygon(cx, cy, r, 3, 30)

class StippleHexagonsPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Hexagons"
    def _generate_shape(self, cx: float, cy: float, r: float): return generate_polygon(cx, cy, r, 6, 0)

class StippleStarsPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Stars"
    def _generate_shape(self, cx: float, cy: float, r: float): 
        # Hacky star
        return generate_polygon(cx, cy, r, 5, 0) + generate_polygon(cx, cy, r*0.5, 5, 36)

class StippleCrossesPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Crosses"
    def _generate_shape(self, cx: float, cy: float, r: float):
        return [(cx-r, cy), (cx+r, cy), (cx, cy), (cx, cy-r), (cx, cy+r)]
        
class StippleLinesPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Lines"
    def _generate_shape(self, cx: float, cy: float, r: float): return [(cx-r, cy-r), (cx+r, cy+r)]

class StippleVariableCirclesPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Variable Circles"
    def _generate_shape(self, cx: float, cy: float, r: float): 
        # Add random jitter to radius
        return generate_circle(cx, cy, r * (0.5 + self._rng.random()), 8)

class StippleVariableSquaresPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Variable Squares"
    def _generate_shape(self, cx: float, cy: float, r: float): 
        return generate_polygon(cx, cy, r * (0.5 + self._rng.random()), 4, self._rng.random()*90)

class StippleChaosPFM(BaseStipplePFM):
    @property
    def name(self) -> str: return "Stipple Chaos"
    def _generate_shape(self, cx: float, cy: float, r: float): 
        pts = int(self._rng.random() * 4) + 3
        return generate_polygon(cx, cy, r, pts, self._rng.random()*360)

STIPPLE_EXTRAS = [
    StippleCirclesPFM(), StippleSquaresPFM(), StippleTrianglesPFM(), 
    StippleHexagonsPFM(), StippleStarsPFM(), StippleCrossesPFM(),
    StippleLinesPFM(), StippleVariableCirclesPFM(), StippleVariableSquaresPFM(), StippleChaosPFM()
]
